src/version.rs

```diff
@@ -194,6 +249,72 @@ impl Version {
                           &[&yanked, &self.id]));
         Ok(())
     }
+
+    pub fn store_build_info(&self,
+                            conn: &GenericConnection,
+                            info: upload::VersionBuildInfo,
+                            krate: &Crate) -> CargoResult<()> {
+
+        // Verify specified Rust version will parse before doing any inserting
+        let channel_version = info.channel_version()?;
+
+        // Future improvement: allow overwriting an existing row, in which case
+        // the ON CONFLICT DO should set `passed` instead of doing nothing.
+        let inserted = conn.execute("INSERT INTO build_info \
+                          (version_id, rust_version, target, passed) \
+                      VALUES ($1, $2, $3, $4) \
+                      ON CONFLICT (version_id, rust_version, target) DO NOTHING",
+            &[&self.id, &info.rust_version, &info.target, &info.passed])?;
+
+        if inserted == 0 {
+            return Err(human(format!(
+                "Build info already specified for {} v{} with {} and target {}",
```

Comment:
I'd guess that overwriting is desired, right? That way if you have a flaky test and rerun the builds it'll appear working?